llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1180,6 +1180,82 @@ Operation *ReverseOp::getTiledImplementation(OpBuilder &builder,
   return tiledRevOp;
 }
 
+//===----------------------------------------------------------------------===//
+// TopkOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult TopkOp::verify() {
+  Operation *op = getOperation();
+  if (getNumInputs() != 2) {
+    return op->emitOpError("expected two input operands");
+  }
+  if (getNumOutputs() != 2) {
+    return op->emitOpError("expected two output operands");
+  }
+  if (!inputValues().getType().isa<ShapedType>() ||
+      !inputIndices().getType().isa<ShapedType>()) {
+    return op->emitOpError("expected input element types to be shaped");
+  }
+  if (!outputValues().getType().isa<ShapedType>() ||
+      !outputIndices().getType().isa<ShapedType>()) {
+    return op->emitOpError("expected output element types to be shaped");
+  }
+  // Ensure input/output values types match
+  ShapedType inputValuesType = inputValues().getType().cast<ShapedType>();
+  ShapedType outputValuesType = outputValues().getType().cast<ShapedType>();
+  if (inputValuesType.getElementType() != outputValuesType.getElementType()) {
+    return op->emitOpError("expected input/output value types to be identical");
+  }
+  // Indices must be index or float
+  ShapedType inputIndicesType = inputIndices().getType().cast<ShapedType>();
+  ShapedType outputIndicesType = outputIndices().getType().cast<ShapedType>();
+  if (!inputIndicesType.getElementType().isIntOrIndex() ||
+      !outputIndicesType.getElementType().isIntOrIndex()) {
+    return op->emitOpError(
+        "expected input/output indices types to be int or index");
+  }
+  // Ranks must match
+  if (inputValuesType.getRank() != outputValuesType.getRank() ||
+      inputIndicesType.getRank() != outputIndicesType.getRank()) {
+    return op->emitOpError("expected input/output to have the same rank");
+  }
+  // Input indicies/values and output indices/values should have the same shape
+  if (llvm::any_of(
+          llvm::zip(inputValuesType.getShape(), inputIndicesType.getShape()),
+          [](std::tuple<int64_t, int64_t> s) {
+            return std::get<0>(s) != ShapedType::kDynamicSize &&
```

Comment:
Why do we have a special case for dynamic size?

Reply:
I copied this pattern from other ops, so there may be some context I'm missing, but from my own tril/error, if an op is dynamic, the shape isn't known at compile time and so `getShape()` will return `-1` for that dimension which I'm guessing means "unknown"? 

Didn't really make sense to compare explicitly unknown dimensions.

Reply:
You could have a case where one of the dims is statically known and other is dynamic before some canonicalization kicks in to make them the same shape. If we dont have this then you will keep getting verification errors.

Reply:
if both dimensions are dynamic I think this should be valid and it up to user to make sure the dynamic dimensions are actually equal. This is how other linalg ops work.

Reply:
> You could have a case where one of the dims is statically known and other is dynamic before some canonicalization kicks in to make them the same shape. If we dont have this then you will keep getting verification errors.

ok, interesting

Reply:
not sure that buys us much. I'd rather allow dimensions that are equal to allow one of them to be static and other to be dynamic and then canonicalize the dynamic one to be the same static. AFAIK that is how Linalg ops work as well.

Reply:
So is this check ok as is? 

Re mahesh: if one dim is dynamic and the other is static, this check will still pass since it tests if both dimensions are't static before actually equating them. Its not possible to compare one dynamic and one static dimension as I mentioned they dynamic is always `-1`.